kotlin-checks-test-sources/src/main/kotlin/checks/IndexedAccessCheckSample.kt

```diff
@@ -0,0 +1,39 @@
+package checks
+
+class IndexedAccessCheckSample {
+
+    fun withoutIndexedAccessors(list: MutableList<Int>, map: MutableMap<String, Int>, grid: Grid) {
+        list.get(1) // Noncompliant {{Replace function call with indexed accessor.}}
+//           ^^^^^^
+        list.set(1, 42) // Noncompliant {{Replace function call with indexed accessor.}}
+//           ^^^^^^^^^^
+        map.get("b") // Noncompliant {{Replace function call with indexed accessor.}}
+        map.set("b", 42) // Noncompliant {{Replace function call with indexed accessor.}}
+//          ^^^^^^^^^^^^
+        grid.get(1, 2) // Noncompliant {{Replace function call with indexed accessor.}}
+        grid.set(1, 2, 42) // Noncompliant {{Replace function call with indexed accessor.}}
+    }
+
+    fun withIndexedAccessors(lisp: Lisp<Int>, list: MutableList<Int>, map: MutableMap<String, Int>, grid: Grid) {
+        lisp.get(1) // Compliant, not an operator
+        list[1] // Compliant
+        list[1] = 42 // Compliant
+        map["b"] // Compliant
+        map["b"] = 42 // Compliant
+        grid[1, 2] // Compliant
+        grid[1, 2] = 42 // Compliant
+        list.getOrNull(2) // Complaint, not an operator
+        list.getOrElse(3) {42} // Complaint, not an operator
+        map.getValue("a") // Complaint, not an operator
+        map.getOrElse("c") {42} // Complaint, not an operator
+    }
+}
+
+interface Grid {
+    operator fun get(row: Int, column: Int): Int
+    operator fun set(row: Int, column: Int, value: Int)
+}
+
+interface Lisp<T> {
+    /*operator*/ fun get(index: Int)
+}
```

Comment:
We should have some test cases about access to indexed elements of it's own class or parent class using `get(index)`, `this.get(index)`, `this@ChildClass.get(index)`, `super.get(index)`. I notice that we have a false positive about `super` and we should make our decision explicit about `this`.

Note: the rule S6519 is also impacted by operator overloading limitation, perhaps some logic related to `super` and `this` could be put in common code. The logic is [here](https://github.com/SonarSource/sonar-kotlin/blob/42ccc6518f9b01052472695d91306f167bfd84cd/sonar-kotlin-plugin/src/main/java/org/sonarsource/kotlin/checks/EqualsMethodUsageCheck.kt#L45)

```kotlin
open class ParentClass {
    private val _array = mutableListOf<String>()

    open operator fun get(index: Int): String {
        return _array[index]
    }

    open operator fun set(index: Int, value: String) {
        _array[index] = value
    }

    val size: Int get() = _array.size
}

class ChildClass : ParentClass() {

    override operator fun get(index: Int): String {
        // The following code does not compile: return super[index]
        return super.get(index) // Noncompliant but false positive, calling "get" here should be compliant
    }

    override operator fun set(index: Int, value: String) {
        // The following code does not compile: super[index] = value
        super.set(index, value) // Noncompliant but false positive, calling "set" here should be compliant
    }


    fun getOrEmpty(index: Int): String =
        if (index in 0 until size)
            get(index) // Noncompliant and it's not possible to simplify with "[index]", only with "this[index]"
                       // and I don't have a strong feeling that we should enforce this case.
        else
            ""

    fun setIfExist(index: Int, value: String) {
        if (index in 0 until size) {
            set(index, value) // Noncompliant and it's not possible to simplify with "[index] = value", only with "this[index] = value"
                              // and I don't have a strong feeling that we should enforce this case.
        }
    }

    inner class InnerClass {
        operator fun get(index: Int): String {
            return this@ChildClass.get(index) // Noncompliant but should we really enforce this delegating case?
        }

        operator fun set(index: Int, value: String) {
            this@ChildClass.set(index, value) // Noncompliant but should we really enforce this delegating case?
        }
    }
}
```

Reply:
Good point about the `super` and `this`! Thanks for the test cases. The logic should be: "if a dot expression (i.e., with an lhs receiver) uses a get / set operator function, it should be replaced with an indexed accessor". According to that, I fixed two of the three suggested corner cases:

- `super.get(index)` -> follows the logic, but `super[index]` is not allowed, therefore FP. Fixed.
- `get(index)` -> Does not follow the logic, because this is no dot expression, therefore FP. I.e., `[index]` is not allowed, while `this[index]` is not a simplification of the original expression. Fixed.
- `return this@ChildClass.get(index)` -> does follow the logic and `return this@ChildClass[index]` is a simplification. Therefore not an FP in my eyes. Not fixed.

Would you agree with that?